source/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.cc

```diff
@@ -0,0 +1,70 @@
+#include "extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.h"
+
+#include <chrono>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "common/common/assert.h"
+
+#include "extensions/filters/http/adaptive_concurrency/concurrency_controller/concurrency_controller.h"
+#include "extensions/filters/http/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdaptiveConcurrency {
+
+AdaptiveConcurrencyFilterConfig::AdaptiveConcurrencyFilterConfig(
+    const envoy::config::filter::http::adaptive_concurrency::v2alpha::AdaptiveConcurrency&,
+    Runtime::Loader& runtime, std::string stats_prefix, Stats::Scope& scope,
+    TimeSource& time_source)
+    : runtime_(runtime), stats_prefix_(std::move(stats_prefix)), scope_(scope),
+      time_source_(time_source) {
+  // TODO (tonya11en): Remove these noop calls when stats/runtime values are
+  // implemented.
+  //
+  // Calling for test coverage.
+  runtime_.snapshot();
+  scope_.constSymbolTable();
+}
+
+AdaptiveConcurrencyFilter::AdaptiveConcurrencyFilter(
+    AdaptiveConcurrencyFilterConfigSharedPtr config, ConcurrencyControllerSharedPtr controller)
+    : config_(std::move(config)), controller_(std::move(controller)) {}
+
+AdaptiveConcurrencyFilter::~AdaptiveConcurrencyFilter() = default;
+
+Http::FilterHeadersStatus AdaptiveConcurrencyFilter::decodeHeaders(Http::HeaderMap&, bool) {
+  if (!forwarding_action_) {
+    forwarding_action_ = std::make_unique<ConcurrencyController::RequestForwardingAction>(
+        controller_->forwardingDecision());
+  }
+
+  if (*forwarding_action_ == ConcurrencyController::RequestForwardingAction::Block) {
+    // TODO (tonya11en): Remove filler words.
+    decoder_callbacks_->sendLocalReply(Http::Code::ServiceUnavailable, "filler words", nullptr,
+                                       absl::nullopt, "more filler words");
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+void AdaptiveConcurrencyFilter::decodeComplete() {
```

Comment:
In the decode path, you need to always make your forwarding decision in decodeHeaders(), regardless of when the request ends. As such, you don't need this decodeComplete() function or the forwarding_action_ pointer. You can just do it all in decodeHeaders(). Please also make sure you have tests that cover various types of requests including headers only, headers with body, headers with body and trailers.